### PR TITLE
Adding LDA Transform

### PR DESCRIPTION
**Pull request by ganik**
issue #373

### CODE REVIEW

**Comment by ganik** [comment]
Step 1: adding LDA Native code

**Comment by ganik** [comment]
test OSX10.13 Debug

**Comment by Ivanidzo4ka** [comment]
> using System;

add .net foundation header

---
Refers to: src/Microsoft.ML.Transforms/Text/LdaSingleBox.cs:1 in 179d6cf. [](commit_id = 179d6cfd2c3ba18a732b82526e05406369ba0937, deletion_comment = False)

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 1]
```diff
@@ -0,0 +1,978 @@
+﻿//------------------------------------------------------------------------------
+// <copyright company="Microsoft Corporation">
+//     Copyright (c) Microsoft Corporation. All rights reserved.
+// </copyright>
+//------------------------------------------------------------------------------
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: change to .net foundation header

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 2]
```diff
@@ -0,0 +1,978 @@
...
+using Microsoft.ML.Runtime.TextAnalytics;
+
+[assembly: LoadableClass(typeof(LdaTransform), typeof(LdaTransform.Arguments), typeof(SignatureDataTransform),
+    LdaTransform.UserName, LdaTransform.LoaderSignature, LdaTransform.ShortName, DocName = "transform/LdaTransform.md")]
+
+[assembly: LoadableClass(typeof(LdaTransform), null, typeof(SignatureLoadDataTransform),
+    LdaTransform.UserName, LdaTransform.LoaderSignature)]
+
+/*
+ * REVIEW taifengw(tfinley): So this would be my starting point. Much of the plumbing is already
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >taifengw(tfinley) [](start = 10, length = 17)

strip all aliases from code

File: src/Native/LDANative/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    lda_engine_export.cpp
+    light_doc_sampler.cpp
+    light_hash_map.cpp
+    model_block.cpp
+    utils.cpp
+)
+
+if(WIN32)
+else()
+    set_property(SOURCE segment.cpp APPEND_STRING PROPERTY COMPILE_FLAGS " -msse4.1")
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/CMakeLists.txt]: >segment.cpp [](start = 24, length = 11)

you don't have to copypaste this line from FastTree cmake :)

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 4]
```diff
@@ -0,0 +1,976 @@
...
+
+[assembly: LoadableClass(typeof(LdaTransform), typeof(LdaTransform.Arguments), typeof(SignatureDataTransform),
+    LdaTransform.UserName, LdaTransform.LoaderSignature, LdaTransform.ShortName, DocName = "transform/LdaTransform.md")]
+
+[assembly: LoadableClass(typeof(LdaTransform), null, typeof(SignatureLoadDataTransform),
+    LdaTransform.UserName, LdaTransform.LoaderSignature)]
+
+/*
+ * REVIEW: So this would be my starting point. Much of the plumbing is already
+ * done, except for the parts indicated by "REVIEW taifengw(tfinley):" in comments. I have not
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >REVIEW taifengw(tfinley) [](start = 44, length = 24)

weird...

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 5]
```diff
@@ -0,0 +1,976 @@
...
+            [TlcModule.SweepableDiscreteParam("NumIterations", new object[] { 100, 200, 300, 400 })]
+            public int NumIterations = 200;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Compute log likelihood over local dataset on this iteration interval", ShortName = "llInterval")]
+            public int LikelihoodInterval = 5;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The threshold of maximum count of tokens per doc", ShortName = "maxNumToken", SortOrder = 50)]
+            public int NumMaxDocToken = 512;
+
+            // REVIEW wenhanw: Should change the default when multi-threading support is optimized.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >wenhanw [](start = 22, length = 7)

clean it

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs (hunk 4 above)
Comment (TomFinley) [reply src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: Uh right. Please just remove this comment. It was back when Taifeng was writing the transform, I furnished him with a "stub" of a transform implementation, with comments on what he would need to write where. This is no longer relevant.

---
In reply to: [197308219](https://github.com/dotnet/machinelearning/pull/377#discussion_r197308219) [](ancestors = 197308219)

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 6]
```diff
@@ -0,0 +1,976 @@
...
+    public sealed class LdaTransform : OneToOneTransformBase
+    {
+        public sealed class Arguments : TransformInputBase
+        {
+            [Argument(ArgumentType.Multiple | ArgumentType.Required, HelpText = "New column definition(s) (optional form: name:srcs)", ShortName = "col", SortOrder = 49)]
+            public Column[] Column;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The number of topics in the LDA", SortOrder = 50)]
+            [TGUI(SuggestedSweeps = "20,40,100,200")]
+            [TlcModule.SweepableDiscreteParam("NumTopic", new object[] { 20, 40, 100, 200 })]
```
Comment (TomFinley) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >object [](start = 62, length = 6)

Do these really have to be `object`? `SweepableDiscreteParam` isn't a generic type?

File: src/Native/LDANative/alias_multinomial_rng_int.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+#include "alias_multinomial_rng_int.hpp"
+#include "rand_int_rng.h"
+#include <ctime>
+#include <list>
+#include <algorithm>
+#include <iostream>
+
+namespace wood
+{
+	AliasMultinomialRNGInt::AliasMultinomialRNGInt()
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/alias_multinomial_rng_int.cpp]: >	 [](start = 0, length = 1)

I know what you just porting files from one repo to another, but....
Can we unify tab vs space story?
some files have tabs, some spaces

File: src/Native/LDANative/hybrid_alias_map.h [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	};
+
+	inline int32_t hybrid_alias_map::size() const
+	{
+		return size_;
+	}
+
+	inline int32_t hybrid_alias_map::next(wood::xorshift_rng &rng, int32_t beta_height, float beta_mass, std::vector<wood::alias_k_v> &beta_k_v, bool debug)
+	{
+		//note(taifengw), here we will set those unseen words' topic to 0. logicall we could set it to random as well.
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/hybrid_alias_map.h]: >taifengw [](start = 9, length = 8)

clean

File: src/Native/CMakeLists.txt [hunk 9]
```diff
@@ -179,4 +179,5 @@ function(install_library_and_symbols targetName)
 endfunction()
 
 add_subdirectory(CpuMathNative)
-add_subdirectory(FastTreeNative)
\ No newline at end of file
+add_subdirectory(FastTreeNative)
+add_subdirectory(LDANative)
```
Comment (TomFinley) [inline src/Native/CMakeLists.txt]: >(LDA [](start = 16, length = 4)

Can we change the name of the directory from `LDANative` to `LdaNative`, to be consistent with other usage elsewhere (the build artifact being `LdaNative.dll`, and so on)?

**Comment by TomFinley** [comment]
No hope of using [https://github.com/Microsoft/lightlda](https://github.com/Microsoft/lightlda) directly?

File: src/Native/LDANative/data_block.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        return (int)data_length;
+    }
+
+    std::shared_ptr<LDADocument> LDADataBlock::GetOneDoc(int32_t index) const
+    {
+        std::shared_ptr<LDADocument> returned_ptr(
+            new LDADocument(documents_buffer_ + offset_buffer_[index],
+            documents_buffer_ + offset_buffer_[index + 1]
+            )
+            );
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/data_block.cpp]: I'm a bit confuse with this one

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs (hunk 1 above)
Comment (ganik) [reply src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: done

---
In reply to: [196948540](https://github.com/dotnet/machinelearning/pull/377#discussion_r196948540) [](ancestors = 196948540)

File: src/Native/LDANative/alias_multinomial_rng_int.hpp [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+                        proportion_int_[id]--;
+                        ++i;
+                    }
+                    id = (id + 1) % n_;
+                }
+            }
+
+            if (mass_sum < mass_int_)
+            {
+                //Todo: Jiyuan, is this data type safe? more is int and mass_sum is in64
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/alias_multinomial_rng_int.hpp]: >Jiyuan [](start = 24, length = 6)

I would strip name, and probably rephrase whole comment

File: src/Native/LDANative/alias_multinomial_rng_int.cpp [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        if (internal_memory_ != nullptr)
+        {
+            delete[]internal_memory_;
+        }
+    }
+    
+    int32_t AliasMultinomialRNGInt::Next(xorshift_rng& rng, std::vector<alias_k_v>& alias_kv)
+    {
+        // NOTE(jiyuan): stl uniform_real_distribution generates the highest quality random numbers
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/alias_multinomial_rng_int.cpp]: >jiyuan [](start = 16, length = 6)

remove aliases

File: src/Native/LDANative/alias_multinomial_rng_int.cpp [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        // CHECK(idx < K.size()) << "sample = " << sample << ",a_int_ = " << a_int_ << ",idx=" << idx;
+        // the following code is equivalent to 
+        // return sample < V_[idx] ? idx : K_[idx];
+        // but faster, see
+        // http://stackoverflow.com/questions/6754454/speed-difference-between-if-else-and-ternary-operator-in-c
+        int m = -(sample < alias_kv[idx].v_);
+        return (idx & m) | (alias_kv[idx].k_ & ~m);
+        // return sample < V[idx] ? idx : K[idx];
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/alias_multinomial_rng_int.cpp]: >// return sample < V[idx] ? idx : K[idx]; [](start = 8, length = 41)

should we remove this line?

File: src/Native/LDANative/data_block.cpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+        offset_buffer_[0] = 0;
+    }
+
+
+    //term_id, term_freq, term_num
+    int LDADataBlock::Add(int32_t* term_id, int32_t* term_freq, int32_t term_num)
+    {
+        int64_t data_length = 1;
+        // offset_buffer_[index_document_] = used_size_;
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/data_block.cpp]: can we remove this line?

File: src/Native/LDANative/data_block.cpp [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        offset_buffer_[index_document_] = used_size_;
+        has_read_ = true;
+
+        return (int)data_length;
+    }
+
+    int LDADataBlock::AddDense(int32_t* term_freq, int32_t term_num)
+    {
+        int64_t data_length = 1;
+        // offset_buffer_[index_document_] = used_size_;
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/data_block.cpp]: >// offset_buffer_[index_document_] = used_size_; [](start = 8, length = 48)

can we remove this line?

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs (hunk 1 above)
Comment (ganik) [reply src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: #Resolved

File: src/Native/LDANative/hybrid_alias_map.h [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+        int32_t *kv_;
+        int32_t *idx_;
+        int32_t height_;
+        int32_t capacity_;
+        int32_t size_;
+
+        float mass_;
+        float n_kw_mass_;
+        float beta_mass_;
+        // static std::vector<wood::alias_k_v> beta_kv_;
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/hybrid_alias_map.h]: >// static std::vector<wood::alias_k_v> beta_kv_; [](start = 8, length = 48)

remove this.

File: src/Native/LDANative/CMakeLists.txt (hunk 3 above)
Comment (ganik) [reply src/Native/LDANative/CMakeLists.txt]: #Resolved

---
In reply to: [196948917](https://github.com/dotnet/machinelearning/pull/377#discussion_r196948917) [](ancestors = 196948917)

File: src/Native/LDANative/lda_engine.cpp [hunk 17]
```diff
@@ -0,0 +1,1103 @@
...
+
+        int thread_id = atomic_stats_->thread_counter_++;
+        std::vector<std::pair<int, double>> llcontainer;
+        // Set core affinity which helps performance improvement
+#ifdef _MSC_VER
+        long long maskLL = 0;
+        maskLL |= (1LL << (thread_id));
+        DWORD_PTR mask = maskLL;
+        SetThreadAffinityMask(GetCurrentThread(), mask);
+#elif !defined(__APPLE__)
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >!defined(__APPLE__) [](start = 6, length = 19)

what about __APPLE__?

File: src/Native/LDANative/lda_engine.cpp [hunk 18]
```diff
@@ -0,0 +1,1103 @@
...
+
+        document_buffer_ = new int32_t*[num_threads_];
+        for (int i = 0; i < num_threads_; i++)
+            document_buffer_[i] = new int32_t[maxDocToken_ * 2 + 1];
+
+        likelihood_in_iter_ = nullptr;
+        beta_sum_ = beta_ * V_;
+
+        //set up some trainig parameter, e.g. topicNum, docNum, 
+        //model_block_.Read(meta_name);
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: do we need this one?

File: src/Native/LDANative/lda_engine.cpp [hunk 19]
```diff
@@ -0,0 +1,1103 @@
...
+
+            samplerQueue_->push(i);
+        }
+        CheckFunction(0, tmDebug, "create samplers", false);
+        return true;
+    }
+
+    void LdaEngine::InitializeBeforeTest()
+    {
+        // TODO(jiyuan):
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >jiyuan [](start = 16, length = 6)

strip alias

File: src/Native/LDANative/lda_engine.cpp [hunk 20]
```diff
@@ -0,0 +1,1103 @@
...
+                    proportion[k] = beta_ / (global_summary_row_[k] + beta_sum_);
+                    beta_mass_ += proportion[k];
+                }
+
+                alias_rng_int_.SetProportionMass(proportion, beta_mass_, beta_k_v_, &beta_height_, sampler_.rng());
+            }
+            process_barrier_->wait();
+            CheckFunction(thread_id, tmDebug, "built alias table dense - in function train_thread");
+
+            // 2,  build alias table for the sparse term
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >, [](start = 16, length = 1)

so in this code they have
1,
2,
3.
4.
which freaks me out, can we use one thing and I would prefer dot for it?

File: src/Native/LDANative/lda_engine.cpp [hunk 21]
```diff
@@ -0,0 +1,1103 @@
...
+#endif
+        process_barrier_->wait();
+
+        //// Each thread builds a portion of word-topic table. We do this way because each word-topic row 
+        //// has a thread-specific buffer for rehashing
+        LightDocSampler &sampler_ = *(samplers_[thread_id]);
+        sampler_.AdaptAlphaSum(false);
+
+        //sampler_.build_word_topic_table(thread_id, num_threads_, model_block_);
+        //process_barrier_->wait();
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: is it needed?

File: src/Native/LDANative/lda_engine.cpp [hunk 22]
```diff
@@ -0,0 +1,1103 @@
...
+        static_cast<double>(token_num) / (init_end - init_start));*/
+
+        process_barrier_->wait();
+
+        /*if (thread_id == 0)
+        {
+        printf("Global Init OK");
+        printf("Start aggreating word_topic_delta, thread = %d\n", thread_id);
+        }
+        process_barrier_->wait();*/
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: do we need all this commented stuff?

File: src/Native/LDANative/lda_engine.cpp [hunk 23]
```diff
@@ -0,0 +1,1103 @@
...
+        if (numThread > 0)
+        {
+            num_threads_ = numThread;
+        }
+        else
+        {
+            unsigned int uNumCPU = std::thread::hardware_concurrency();
+            num_threads_ = std::max(1, (int)(uNumCPU - 2));
+        }
+        printf("using %d thread(s) to do train/test\n", num_threads_);
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >printf [](start = 8, length = 6)

can you create issue to strip out all std cout and printf from this code and replace them with Message function call?

File: src/Native/LDANative/lda_engine.cpp [hunk 24]
```diff
@@ -0,0 +1,1103 @@
...
+        process_barrier_->wait();
+        sampler_.build_alias_table(word_range_for_each_thread_[thread_id], word_range_for_each_thread_[thread_id + 1], thread_id);
+        process_barrier_->wait();
+
+        /*double alias_end = lda::get_time();
+        if (thread_id == 0)
+        {
+        std::cout << "Elapsed time for building alias table: " << (alias_end - alias_start) << std::endl;
+        }
+        process_barrier_->wait();*/
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: is it required?

File: src/Native/LDANative/lda_engine.cpp [hunk 25]
```diff
@@ -0,0 +1,1103 @@
...
+
+    int LdaEngine::FeedInData(int* term_id, int* term_freq, int32_t term_num, int32_t vocab_size)
+    {
+        // bool b_ret = true;
+        if (V_ == 0) //number vocab could be set in allocating model memory function
+            V_ = vocab_size;
+
+        //data_block represent for one doc
+        return data_block_->Add(term_id, term_freq, term_num);
+        // return b_ret;
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >// return b_ret; [](start = 8, length = 16)

can we clean it? and commented definition above. Same for FeedInDataDense function

File: src/Native/LDANative/lda_engine.cpp [hunk 26]
```diff
@@ -0,0 +1,1103 @@
...
+        assert(data_length <= maxDocToken_ * 2 + 1);
+
+        if (reset)
+        {
+            // restart the rng seeds, so that we always get consistent result for the same input
+            rng_.restart();
+            sampler.rng_restart();
+        }
+
+        // NOTE(jiyuan): in multi-threaded implementation, the dynamic memory allocation
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >jiyuan [](start = 16, length = 6)

remove alias

File: src/Native/LDANative/lda_engine.cpp [hunk 27]
```diff
@@ -0,0 +1,1103 @@
...
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >jiyuan [](start = 16, length = 6)

remove alias

File: src/Native/LDANative/lda_engine.cpp [hunk 28]
```diff
@@ -0,0 +1,1103 @@
...
+        return true;
+    }
+
+    //function to support dumping the topic_model model file
+    void LdaEngine::GetWordTopic(int32_t wordId, int32_t* pTopic, int32_t* pProb, int32_t& length)
+    {
+        //cap the topic number here according to inpassed value of length
+        int lengthCap = length;
+
+        // NOTE(jiyuan): we MUST check whether the word-topic row is empty before get its value
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >jiyuan [](start = 16, length = 6)

alias

File: src/Native/LDANative/lda_engine.cpp [hunk 29]
```diff
@@ -0,0 +1,1103 @@
...
+                if (length >= lengthCap)
+                    break;
+            }
+        }
+    }
+
+    // Compare by frequencies in descending order.
+    bool CompareTerms(const std::pair<int, int> &term1, const std::pair<int, int> &term2)
+    {
+        // REVIEW wenhanw(yaeld): consider changing this to impose a total order, since quicksort is not stable.
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >wenhanw(yaeld) [](start = 18, length = 14)

alias

File: src/Native/LDANative/lda_engine.cpp [hunk 30]
```diff
@@ -0,0 +1,1103 @@
...
+        {
+            pWords[i] = allTermsVec[i].first;
+            pProb[i] = (((float)(allTermsVec[i].second)) + beta_) / (sumFreq + beta_ * V_);
+        }
+    }
+
+    //function to support loading the topic_model model file
+    void LdaEngine::SetWordTopic(int32_t wordId, int32_t* pTopic, int32_t* pProb, int32_t length)
+    {
+        //Note: taifengw(jinhui) whether we should really use the "true" here
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >taifengw(jinhui) [](start = 16, length = 16)

alias

File: src/Native/LDANative/lda_engine.cpp [hunk 31]
```diff
@@ -0,0 +1,1103 @@
...
+        for (int i = 0; i < length; ++i)
+        {
+            global_word_topic_table_[wordId].inc(pTopic[i], pProb[i]);
+            global_summary_row_[pTopic[i]] += pProb[i];
+        }
+    }
+
+    void LdaEngine::GetModelStat(int64_t &memBlockSize, int64_t &aliasMemBlockSize)
+    {
+        //Note: taifengw, get the model's value at the end of training stage. try to save these two numbers to disk file
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: >//Note: taifengw, get the model's value at the end of training stage. try to save these two numbers to disk file [](start = 8, length = 112)

remove it probably, or rephrase

File: src/Native/LDANative/lda_engine.cpp [hunk 32]
```diff
@@ -0,0 +1,1103 @@
...
+        for (int w = 0; w < V_; ++w)
+        {
+            int nonzero_num = global_word_topic_table_[w].nonzero_num();
+            if (nonzero_num)
+            {
+                wt_stream << w;
+                for (int t = 0; t < K_; ++t)
+                {
+                    // if (word_topic_table_[w * K_ + t])
+                    // if (word_topic_table_[(int64_t)w * K_ + t])
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.cpp]: remove it

File: src/Native/LDANative/lda_engine.hpp [hunk 33]
```diff
@@ -0,0 +1,146 @@
...
+
+        //IO, data
+        bool ClearData();  //for clean up training data
+        bool ClearModel(); //for testing purpose, before calling SetWordTopic, please clear the old model
+
+        int FeedInData(int* term_id, int* term_freq, int32_t term_num, int32_t vocab_size);
+        int FeedInDataDense(int* term_freq, int32_t term_num, int32_t vocab_size);
+
+        //IO, model 
+        // NOTE(jiyuan): assume pTopic and pProb are allocated outside the function
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.hpp]: >jiyuan [](start = 16, length = 6)

alias

File: src/Native/LDANative/lda_engine.hpp [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+
+
+        void InitializeBeforeTest();
+        bool InitializeBeforeTrain();
+        void AllocateDataMemory(int num_document, int64_t corpus_size);
+        void AllocateModelMemory(const LDADataBlock* data_block); //in this case, model memory is allocated according to the datablock;
+        void AllocateModelMemory(int num_vocabs, int num_topics, int64_t nonzero_num);
+        void AllocateModelMemory(int num_vocabs, int num_topics, int64_t mem_block_size, int64_t alias_mem_block_size);
+        //void SetNonzeroNum(int32_t word_id, int32_t nonzero_num);
+        //void FinallizeNonzero();
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine.hpp]: remove it

File: src/Native/LDANative/lda_engine_export.cpp [hunk 35]
```diff
@@ -0,0 +1,109 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "lda_engine.hpp"
+
+/// This file use to expose public API to be consumed by TLC.
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/lda_engine_export.cpp]: >TLC [](start = 57, length = 3)

ML.Net

File: src/Native/LDANative/light_doc_sampler.cpp [hunk 36]
```diff
@@ -0,0 +1,681 @@
...
+    {
+        return p1.first < p2.first;
+    }
+
+    void LightDocSampler::GetDocTopic(LDADocument *doc, int* pTopics, int* pProbs, int32_t& numTopicsMax)
+    {
+        doc_topic_counter_.clear();
+        doc->GetDocTopicCounter(doc_topic_counter_);
+
+        // note(taifengw), do we have to assume this?
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/light_doc_sampler.cpp]: >taifengw [](start = 16, length = 8)

alias

File: src/Native/LDANative/lda_engine.cpp (hunk 17 above)
Comment (ganik) [reply src/Native/LDANative/lda_engine.cpp]: will create issue to track this separately

File: src/Native/LDANative/light_doc_sampler.cpp [hunk 37]
```diff
@@ -0,0 +1,681 @@
...
+
+    void LightDocSampler::Dump(const std::string &dump_name, int32_t lower, int32_t upper)
+    {
+        std::ofstream wt_stream;
+        wt_stream.open(dump_name, std::ios::out);
+        // CHECK(wt_stream.good()) << "Open word_topic_dump file: " << dump_name;
+
+        for (int w = lower; w < upper; ++w)
+        {
+            //taifeng: why not just a serialization of current hybrid_map? do we need to do a search?
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/light_doc_sampler.cpp]: >taifeng [](start = 14, length = 7)

alias

File: src/Native/LDANative/light_hash_map.h [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+
+#define ILLEGAL_BUCKET -1
+
+    class light_hash_map
+    {
+    public:
+
+        // must call set_memory after construction before use
+        light_hash_map();
+        // NOTE(jiyuan): the size of mem_block_ = 2 * capacity_
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/light_hash_map.h]: >jiyuan [](start = 16, length = 6)

alias

File: src/Native/LDANative/model_block.cpp [hunk 39]
```diff
@@ -0,0 +1,467 @@
...
+#pragma warning(suppress: 6386)
+            dict_[i].is_dense_ = 0;
+            dict_[i].is_alias_dense_ = 0;
+        }
+
+        // mem_block_size_ = nonzero_num * 2;        // TODO: power of 2, load factor of 2, key-value of 2
+        mem_block_size_ = 2 * upper_bound(load_factor_ * nonzero_num);
+        alias_mem_block_size_ = nonzero_num * 3; 
+
+        mem_block_ = new int32_t[mem_block_size_]();                // NOTE(jiyuan) : force to initialize the values to be zero
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/model_block.cpp]: >jiyuan [](start = 76, length = 6)

all this aliases

File: src/Native/LDANative/model_block.h [hunk 40]
```diff
@@ -0,0 +1,97 @@
...
+        void Clear();
+        void Init(int32_t num_vocabs, int32_t num_topics);
+        void Init(int32_t num_vocabs, int32_t num_topics, int64_t nonzero_num);
+        void Init(int32_t num_vocabs, int32_t num_topics, int64_t mem_block_size, int64_t alias_mem_block_size);
+
+        void InitFromDataBlock(const LDADataBlock *data_block, int32_t num_vocabs, int32_t num_topics);
+
+        // void InitFromModelFile(const )
+        //void SetNonzeroNum(int32_t word_id, int32_t nonzero_num);
+        //void FinallizeNonzero();
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/model_block.h]: remove them?

File: src/Native/LDANative/rand_int_rng.h [hunk 41]
```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma once
+#include <time.h>
+#include <stdint.h>
+
+namespace wood
```
Comment (Ivanidzo4ka) [inline src/Native/LDANative/rand_int_rng.h]: >wood [](start = 10, length = 4)

what a nice namespace name :)

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 42]
```diff
@@ -0,0 +1,964 @@
...
+            _types = new ColumnType[Infos.Length];
+            _ldas = new LdaState[Infos.Length];
+            _saveText = args.OutputTopicWordSummary;
+            for (int i = 0; i < Infos.Length; i++)
+            {
+                var ex = new ColInfoEx(Host, args.Column[i], args);
+                _exes[i] = ex;
+                _types[i] = new VectorType(NumberType.Float, ex.NumTopic);
+                // C.f. taifengw's shelveset. The following line should be defered in order to pass in numVocab.
+                // _ldas[i] = new LdaState(ex);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: should we remove this?

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 43]
```diff
@@ -0,0 +1,964 @@
...
+            public void AllocateDataMemory(int docNum, long corpusSize)
+            {
+                _ldaTrainer.AllocateDataMemory(docNum, corpusSize);
+            }
+
+            public int FeedTrain(IExceptionContext ectx, ref VBuffer<Double> input)
+            {
+                Contracts.AssertValue(ectx);
+
+                // REVIEW taifengw(tfinley): Input the counts to your trainer here. This
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >taifengw(tfinley) [](start = 26, length = 17)

alias

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 44]
```diff
@@ -0,0 +1,964 @@
...
+                int docSize = 0;
+                int termNum = 0;
+                for (int i = 0; i < src.Count; i++)
+                {
+                    int termFreq = GetFrequency(src.Values[i]);
+                    if (termFreq < 0)
+                    {
+                        // REVIEW: Should this log a warning message? And what should it produce?
+                        // It currently produces a vbuffer of all NA values.
+                        // REVIEW shonk: Need a utility method to do this...
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >shonk [](start = 34, length = 5)

alias

**Comment by ganik** [comment]
test OSX10.13 Release

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs [hunk 45]
```diff
@@ -0,0 +1,962 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.CommandLine;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.EntryPoints;
```
Comment (ganik) [inline src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: >Microsoft [](start = 6, length = 9)

test

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs (hunk 2 above)
Comment (ganik) [reply src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: d

---
In reply to: [196948610](https://github.com/dotnet/machinelearning/pull/377#discussion_r196948610) [](ancestors = 196948610)

File: src/Native/LDANative/model_block.cpp (hunk 39 above)
Comment (ganik) [reply src/Native/LDANative/model_block.cpp]: #Resolved

Comment (ganik) [reply src/Native/LDANative/model_block.cpp]: #Resolved

---
In reply to: [197899749](https://github.com/dotnet/machinelearning/pull/377#discussion_r197899749) [](ancestors = 197899749)

File: src/Native/LDANative/model_block.h (hunk 40 above)
Comment (ganik) [reply src/Native/LDANative/model_block.h]: #Resolved

**Comment by ganik** [comment]
added unit tests that do that

---
In reply to: [399518917](https://github.com/dotnet/machinelearning/pull/377#issuecomment-399518917) [](ancestors = 399518917)

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs (hunk 6 above)
Comment (ganik) [reply src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: SweepableDiscreteParam  constructor requires that. changing to int for example breaks compilation. To have what you want will need to overload constructor with various types. if you feel strongly about it lets open separate issue to address it

---
In reply to: [197513675](https://github.com/dotnet/machinelearning/pull/377#discussion_r197513675) [](ancestors = 197513675)

**Comment by TomFinley** [comment]
Ah OK. What I meant was their own code, rather than checking it in ourselves... but maybe this is harder, since I think they've probably diverged. Something we should think about though.

---
In reply to: [400106624](https://github.com/dotnet/machinelearning/pull/377#issuecomment-400106624) [](ancestors = 400106624,399518917)

File: src/Microsoft.ML.Transforms/Text/LdaTransform.cs (hunk 6 above)
Comment (TomFinley) [reply src/Microsoft.ML.Transforms/Text/LdaTransform.cs]: That's OK. We'll need to improve this elsewhere I suppose...

---
In reply to: [197954198](https://github.com/dotnet/machinelearning/pull/377#discussion_r197954198) [](ancestors = 197954198,197513675)

**Pull request merged**